modules/db/src/main/scala/gem/dao/StepDao.scala

```diff
@@ -89,21 +90,51 @@ object StepDao {
       }
   }
 
-  def selectAll(oid: Observation.Id): ConnectionIO[List[Step[Instrument]]] =
+  def selectIds(oid: Observation.Id): ConnectionIO[List[Sequence.Id]] =
     sql"""
-      SELECT s.instrument,
-             s.step_type,
-             sg.gcal_lamp,
-             sg.shutter,
-             sc.offset_p,
-             sc.offset_q
-        FROM step s
-             LEFT OUTER JOIN step_gcal sg
-                ON sg.observation_id = s.observation_id AND sg.index = s.index
-             LEFT OUTER JOIN step_science sc
-                ON sc.observation_id = s.observation_id AND sc.index = s.index
-       WHERE s.observation_id = ${oid}
-    ORDER BY s.index
-    """.query[StepKernel].map(_.toStep).list
+          SELECT DISTINCT sequence_id
+            FROM step
+           WHERE observation_id = $oid
+        ORDER BY sequence_id
+       """.query[Sequence.Id].list
 
+  def selectAll(oid: Observation.Id): ConnectionIO[List[Sequence[Step[Instrument]]]] =
+    sql"""
+       SELECT s.sequence_id,
+              s.instrument,
+              s.step_type,
+              sg.gcal_lamp,
+              sg.shutter,
+              sc.offset_p,
+              sc.offset_q
+       FROM step s
+              LEFT OUTER JOIN step_gcal sg
+                 ON sg.sequence_id = s.sequence_id AND sg.index = s.index
+              LEFT OUTER JOIN step_science sc
+                 ON sc.sequence_id = s.sequence_id AND sc.index = s.index
+       WHERE s.observation_id = $oid
+    ORDER BY s.sequence_id, s.index
+    """.query[StepKernel].list.map { (lst: List[StepKernel]) =>
+      lst.groupBy(_.sid).map { case (sid, sks) => Sequence(sid, sks.map(_.toStep)) }.toList
+    }
+
+  def selectAll(sid: Sequence.Id): ConnectionIO[Option[Sequence[Step[Instrument]]]] =
+    sql"""
+       SELECT s.sequence_id,
+              s.instrument,
+              s.step_type,
+              sg.gcal_lamp,
+              sg.shutter,
+              sc.offset_p,
+              sc.offset_q
+       FROM step s
+              LEFT OUTER JOIN step_gcal sg
+                 ON sg.sequence_id = s.sequence_id AND sg.index = s.index
+              LEFT OUTER JOIN step_science sc
+                 ON sc.sequence_id = s.sequence_id AND sc.index = s.index
+       WHERE s.sequence_id = $sid
+    ORDER BY s.index
+    """.query[StepKernel].list.map { (lst: List[StepKernel]) =>
+      lst.headOption.map { sk => Sequence(sk.sid, lst.map(_.toStep)) }
+    }
```

Comment:
These two queries are almost the same.  They differ in the `WHERE` and `ORDER BY` clauses.  Is there a way to share the rest?